### PR TITLE
Blob fetch with no type has an empty content type

### PR DESCRIPTION
**Pull request by dlrobertson**
When fetching a blob with and empty type, the content-type header should be the empty string.

Related to: https://github.com/whatwg/fetch/issues/1436

### CODE REVIEW

File: fetch/api/basic/scheme-blob.sub.any.js [hunk 1]
```diff
@@ -45,4 +45,44 @@ invalidRequestMethods.forEach(function(method) {
 checkKoUrl("blob:not-backed-by-a-blob/", "GET",
            "Fetching [GET] blob:not-backed-by-a-blob/ is KO");
 
+var empty_blob = new Blob([]);
+checkFetchResponse(URL.createObjectURL(empty_blob), "", "", 0,
+                  "Fetching URL.createObjectURL(empty_blob) is OK");
+
+var empty_type_blob = new Blob([], {type: ""});
```
Comment (dlrobertson) [inline fetch/api/basic/scheme-blob.sub.any.js:52]: Note that these tests do not include any content, so a engine with an algorithm that tries to sniff the content type will pass this, but in some cases I _think_ would provide a potentially non-standard content type.

**Review by dlrobertson — COMMENTED** [review]
Note that this patch doesn't include any XHR tests or tests of an iframe with a blob source that does not include a content-type.

File: fetch/api/basic/scheme-blob.sub.any.js [hunk 2]
```diff
@@ -45,4 +45,44 @@ invalidRequestMethods.forEach(function(method) {
 checkKoUrl("blob:not-backed-by-a-blob/", "GET",
            "Fetching [GET] blob:not-backed-by-a-blob/ is KO");
 
+var empty_blob = new Blob([]);
```
Comment (annevk) [inline fetch/api/basic/scheme-blob.sub.any.js:48]: const/let for new code please.

File: fetch/api/basic/scheme-blob.sub.any.js (hunk 1 above)
Comment (annevk) [reply fetch/api/basic/scheme-blob.sub.any.js:52]: Blob content should never be sniffed to determine the response `Content-Type` header though. So maybe we should include a test where the blob starts with a file signature of a JPEG or some such.

**Review by annevk — COMMENTED** [review]
Thanks, these look okay to me, modulo some minor nits.

File: fetch/api/basic/scheme-blob.sub.any.js [hunk 3]
```diff
@@ -45,4 +45,78 @@ invalidRequestMethods.forEach(function(method) {
...
+    return image_blob.arrayBuffer();
+  }).then(function(image_buffer) {
+    let blob = new Blob([image_buffer]);
+    return fetch(URL.createObjectURL(blob)).then(function(resp) {
+      assert_equals(resp.status, 200, "HTTP status is 200");
+      assert_equals(resp.type, "basic", "response type is basic");
+      assert_equals(resp.headers.get("Content-Type"), "", "Content-Type is " + resp.headers.get("Content-Type"));
+    })
+  });
+}, "Blob content is not sniffed for a content type [image/png]");
```
Comment (dlrobertson) [inline fetch/api/basic/scheme-blob.sub.any.js:73]: Used [Response.blob](https://developer.mozilla.org/en-US/docs/Web/API/Response/blob) to get image data. Not sure how we normally do this.

**Review by dlrobertson — COMMENTED** [review]
Added some tests that have content where algorithms that sniff the blob content for a content-type should fail

File: fetch/api/basic/scheme-blob.sub.any.js [hunk 4]
```diff
@@ -45,4 +45,78 @@ invalidRequestMethods.forEach(function(method) {
...
+
+let empty_data_blob = new Blob([], {type: "text/plain"});
+checkFetchResponse(URL.createObjectURL(empty_data_blob), "", "text/plain", 0,
+                  "Fetching URL.createObjectURL(empty_data_blob) is OK");
+
+promise_test(function(test) {
+  return fetch("/images/blue.png").then(function(resp) {
+    return resp.blob();
+  }).then(function(image_blob) {
+    return image_blob.arrayBuffer();
```
Comment (annevk) [inline fetch/api/basic/scheme-blob.sub.any.js:64]: Wouldn't this suffice:
```suggestion
    return resp.arrayBuffer();
```
?

(The alternative would be to peek at the image data and copy the first couple of bytes, but this seems reasonable.)

**Review by annevk — APPROVED** [review]
Thanks, looks good to me!

Another possible thing you could use here are async/await.

File: fetch/api/basic/scheme-blob.sub.any.js (hunk 4 above)
Comment (dlrobertson) [reply fetch/api/basic/scheme-blob.sub.any.js:64]: Neat! Yeah this totally works. For some reason I missed `Response.arrayBuffer`

**Comment by dlrobertson** [comment]
Updated: forgot to `git add` the changes before amending the commit... been spending too much time working with mercurial I guess 😆

File: fetch/api/basic/scheme-blob.sub.any.js [hunk 5]
```diff
@@ -45,4 +45,76 @@ invalidRequestMethods.forEach(function(method) {
...
+                   "Blob content is not sniffed for a content type [text/xml]");
+
+promise_test(function(test) {
+  let blob = new Blob([], {"type": "text/plain"});
+  let slice = blob.slice(8, 25, "\0");
+  return fetch(URL.createObjectURL(slice)).then(function (resp) {
+    assert_equals(resp.status, 200, "HTTP status is 200");
+    assert_equals(resp.type, "basic", "response type is basic");
+    assert_equals(resp.headers.get("Content-Type"), "");
+    assert_equals(resp.headers.get("Content-Length"), "17");
```
Comment (mkruisselbrink) [inline fetch/api/basic/scheme-blob.sub.any.js:85]: this (and the next test) seems wrong. blob is an empty blob, so the slice made out of it is also empty. As such the expected content length would be 0, not some non-zero number. Or alternatively the blob should not be an empty blob.

Comment (dlrobertson) [reply fetch/api/basic/scheme-blob.sub.any.js:85]: Good catch! Originally the input blob for these tests did have data and at some point I must have run tests with an empty blob and forgot to add the content back to them. Created https://github.com/web-platform-tests/wpt/pull/36594 to fix this.

**Pull request merged**